lib/sshutils/scp/scp.go

```diff
@@ -43,110 +41,261 @@ const (
+	// has to be executed on the remove server (handled by Teleport)
+	GetRemoteShellCmd() (string, error)
+}
+
+// FileSystem is an interface that abstracts file system methods used in SCP command functions
+type FileSystem interface {
+	// IsDir returns true if a given file path is a directory
+	IsDir(path string) bool
+	// GetFileInfo returns FileInfo for a given file path
+	GetFileInfo(filePath string) (FileInfo, error)
+	// MkDir creates a directory
+	MkDir(path string, mode int) error
+	// OpenFile opens a file and returns its Reader
+	OpenFile(filePath string) (io.ReadCloser, error)
+	// CreateFile creates a new file
+	CreateFile(filePath string, length uint64) (io.WriteCloser, error)
+	// SetChmod sets file permissions
+	SetChmod(path string, mode int) error
+}
+
+// FileInfo is an API that describes methods that provide file information
+type FileInfo interface {
+	// IsDir returns true if a file is a directory
+	IsDir() bool
+	// ReadDir returns information of directory files
+	ReadDir() ([]FileInfo, error)
+	// GetName returns a file name
+	GetName() string
+	// GetPath returns a file path
+	GetPath() string
+	// GetModePerm returns file permissions
+	GetModePerm() os.FileMode
+	// GetSize returns file size
+	GetSize() int64
+}
+
+// CreateDownloadCommand configures and returns a command used
+// to download a file
+func CreateDownloadCommand(cfg Config) (Command, error) {
+	cfg.Flags.Sink = true
+	cfg.Flags.Source = false
+	cmd, err := CreateCommand(cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-	// terminal is only initialized on the client, for printing the progress
-	Terminal io.Writer
+	return cmd, nil
+}
+
+// CreateUploadCommand configures and returns a command used
+// to upload a file
+func CreateUploadCommand(cfg Config) (Command, error) {
+	cfg.Flags.Sink = false
+	cfg.Flags.Source = true
+	cmd, err := CreateCommand(cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return cmd, nil
+}
+
+// CheckAndSetDefaults checks and sets default values
+func (cfg *Config) CheckAndSetDefaults() error {
+	if cfg.FileSystem == nil {
+		cfg.FileSystem = &localFileSystem{}
+	}
+
+	if cfg.User == "" {
+		return trace.BadParameter("missing User parameter")
+	}
+
+	return nil
+}
+
+// CreateCommand creates and returns a new Command
+func CreateCommand(cfg Config) (Command, error) {
+	err := cfg.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cmd := command{
+		Config: cfg,
+	}
+
+	cmd.Entry = log.WithFields(log.Fields{
+		trace.Component: "SCP",
+		trace.ComponentFields: log.Fields{
+			"LocalAddr":      cfg.Flags.LocalAddr,
+			"RemoteAddr":     cfg.Flags.RemoteAddr,
+			"Target":         cfg.Flags.Target,
+			"User":           cfg.User,
+			"RemoteLocation": cfg.RemoteLocation,
+		},
+	})
+
+	return &cmd, nil
+}
+
+// Command mimics behavior of SCP command line tool
+// to teleport can pretend it launches real scp behind the scenes
+type command struct {
+	Config
+	*log.Entry
```

Comment:
Can you call this variable `log`? Embedded `*log.Entry` like this pollutes autocomplete for me.

Reply:
done